ll2_global_planner/src/ll2_global_planner_nodelet.cpp

```diff
@@ -44,10 +49,178 @@ void Ll2GlobalPlannerNl::loadParams()
+    return;
+  }
+
+  BasicPoint2d starting_point(
+    tf_msg.transform.translation.x, tf_msg.transform.translation.y);
+  const Lanelet& starting_lanelet = getNearestLanelet(starting_point);
+
+  // Plan a route from current vehicle position
+  traffic_rules::TrafficRulesPtr traffic_rules{traffic_rules::TrafficRulesFactory::instance().create(Locations::Germany, Participants::Vehicle)};
+  routing::RoutingGraphUPtr graph = routing::RoutingGraph::build(*lanelet_map_, *traffic_rules);
+  Optional<routing::LaneletPath> shortest_path_opt = graph->shortestPath(starting_lanelet, goal_lanelet);
+  routing::LaneletPath shortest_path;
+
+  if (!shortest_path_opt)
+  {
+    ROS_WARN("Could not find path in lanelet map!");
+    return;
+  }
+
+  shortest_path = shortest_path_opt.value();
+  LaneletSequence continuous_lane = shortest_path.getRemainingLane(shortest_path.begin());
+
+  if (continuous_lane.size() != shortest_path.size())
+  {
+    ROS_WARN("This route contains a lane change which is currently unsupported");
+    return;
+  }
+
+  ROS_INFO("Found a path containing %lu lanelets", shortest_path.size());
+
+  // Convert to autoware waypoints
+  std::vector<autoware_msgs::Waypoint> waypoints;
+
+  // Loop over each lanelet
+  for (auto& lanelet : continuous_lane.lanelets())
+  {
+    const std::string turn_direction = lanelet.attributeOr("turn_direction", "straight");
+    traffic_rules::SpeedLimitInformation speed_limit = traffic_rules->speedLimit(lanelet);
+    ConstLineString3d centerline = lanelet.centerline();
+    int wp_length = centerline.size() - 1;
+
+    // Loop over each centerline point
+    for (int i = 0; i <= wp_length; i++)
+    {
+      auto point = centerline[i];
+
+      autoware_msgs::Waypoint new_wp;
+      new_wp.pose.pose.position.x = point.x();
+      new_wp.pose.pose.position.y = point.y();
+      new_wp.pose.pose.position.z = point.z();
+      new_wp.twist.twist.linear.x = speed_limit.speedLimit.value(); // m/s
+
+      int steering_state = autoware_msgs::WaypointState::STR_STRAIGHT;
+      if (turn_direction.compare("right") == 0)
+      {
+        steering_state = autoware_msgs::WaypointState::STR_RIGHT;
+      }
+      if (turn_direction.compare("left") == 0)
+      {
+        steering_state = autoware_msgs::WaypointState::STR_LEFT;
+      }
+      new_wp.wpstate.steering_state = steering_state;
+
+      waypoints.push_back(new_wp);
+    }
+  }
+
+  int wp_id = 0;
+  bool first_point = true;
+  float smallest_goal_dist = std::numeric_limits<float>::infinity();
+  int smallest_goal_wp_id = 0;
+  autoware_msgs::Waypoint* prev_waypoint = nullptr;
+  std::vector<autoware_msgs::Waypoint> processed_waypoints;
+
+  // Remove duplicates and set waypoint orientations
+  for (autoware_msgs::Waypoint& waypoint : waypoints)
+  {
+    geometry_msgs::Pose& pose = waypoint.pose.pose;
+    geometry_msgs::Pose& prev_pose = prev_waypoint->pose.pose;
+
+    if (!first_point)
+    {
+      // Check for duplicate points
```

Comment:
Are duplicate points a result of subsequent lanelets having a center line point in common? If so should this duplication be avoided in the lanelets centerline to waypoints logic, rather than be removed here?

Reply:
Correct, the very last centerline point of one lanelet may the same point as the first centerline point from the subsequent lanelet. Not sure what you mean, this code is part of the `lanelets centerline to waypoints logic`.

Reply:
I only mean rather than remove duplicates after building up waypoints, when looping through lanelets, avoid adding waypoints if from the beginning of lanelet. 

Probably makes sense to make sure there are no duplicates anyway as a final check after building up waypoints, so I'll leave it up to you.